app/helpers/application_helper.rb

```diff
@@ -1,4 +1,5 @@
 # frozen_string_literal: true
 
 module ApplicationHelper
+  def button_class = "rounded-md bg-indigo-600 px-3 py-2 text-sm font-semibold text-white shadow-sm hover:bg-indigo-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600" # rubocop:disable Layout/LineLength
```

Comment:
@csalvato, is this a common approach? Felt awkward to use helpers to standardize visualization content.

Reply:
I wouldn't put this in a helper, I'd just put it in the view.

A tradeoff of tailwind is really long class names. So I would put it in the view layer, and disable any line length cops there.

Reply:
How do I keep consistency with every button in my multiple views?

Whenever I want to refactor the "standard button" what would I do?

Reply:
There's at least three conventional options here:

1. Define a partial for `button` and whenever you want to render a button, render it using the partial.
2. Create a method in `ApplicationHelper`, or extend `ActionView::Helpers::UrlHelper` to include a method called somethinglike `button_to` (similar to `link_to`) that generates the button with  the classes you want.
3. Slightly less conventional, you can use [view components](https://github.com/ViewComponent/view_component), but I don't have experience here.

It's been a while since I used rails views like this, but I'd probably go the second route, since it's conventional with the `ActionView::Helpers::` API for a lot of rails views...but using partials is more React-like.

Reply:
Hummm... if I understand the options well, we would cover buttons coming from `link_to` but not every button since some will come from `f.submit`.

Reply:
It's been a while since I looked into it, but it should also be possible to overload `f.submit` and `button_for` helpers.

Partials may be easier though.

Reply:
Still struggling with this one. A partial feels like an overkill as I will need to run render every time I want a button. 

Can I use a CSS and add all the Tailwind elements into it? I know this is not how Tailwind was thought to work as each tag defines one clear property. But still, it feels like it would be a ton of changes to keep consistency across my project. 

Overwriting Rails also does not feel great.

Reply:
>I know this is not how Tailwind was thought to work as each tag defines one clear property.

Tailwind emerged in the scene as [web components](https://developer.mozilla.org/en-US/docs/Web/API/Web_Components) was really taking hold in web UI engineering.

Most of the frameworks that existed at the time it was created were component-based - React, Ember, Angular, Vue.

So, Tailwind really emerged in a world where the front end is broken into components. So the long class names are abstracted into a component rather than peppered everywhere you define an element.

There is no real "Rails way" to do components, because Rails was created in a time before web components were a thing. Rails also rejects a lot of the premise of web components.

So the best thing that really fits is using partials, as that's the closest thing to components.

In practice, I've seen vanilla Rails apps do this in a few ways:

1. A bunch of app specific helpers
2. Overloading the helpers that come with Rails
3. Liberally using partials
4. Have a reference of the HTML to copy/paste wherever you use a button.
5. [Using Tailwind's `@apply` to extract classes](https://tailwindcss.com/docs/reusing-styles#extracting-classes-with-apply)

Personally, I think that it's acceptable for a "design system" to simply be HTML that you can copy and paste, and modify wherever you want. This is actually *better* than how most apps that use SCSS are managed, and it's more explicit.

If it were me, I would avoid using `@apply` and use partials liberally.

Reply:
Tailwind with react definitely makes a ton of sense. You create a component and set its visual properties with a better system. So you DRY through the component.

Apply is the approach for what I did with helpers. I don't get why you dislike it as it seems to me as better than repeating it everywhere I have a button (I can only imagine what to do when I want to change the button style). 

I get that as I specialize the visual, I may end up creating a ton of `@apply` elements and it may become a nightmare in itself. Than better to just put on each element. 

The explicit part I also get from the standpoint of someone that could easily really "read" Tailwind.

I am leaning towards either `@apply` or overloading Rails helpers.

Reply:
>I don't get why you dislike it as it seems to me as better than repeating it everywhere I have a button 

I dislike `@apply` because it's no better than using a `.button` css class. So it's basically a slippery slope back to defining a bunch of properties that should go together, and then you're back in the same mess that Tailwind is supposed to fix.

That's why it goes against the Tailwind ethos – it just reintroduces the problem that Tailwind was created to solve.

[This is explained in more detail on the docs page for `@apply` as well](https://tailwindcss.com/docs/reusing-styles#avoiding-premature-abstraction)

If you really do need reusability, I don't see how a partial is worse than `@apply`. Instead of:

``erb
<%= render "button", locals: { text: "Click me!" } %>
```

```html
<button class="btn">Click me!</button>
```

Reply:
I get the reasoning. 

I rolled back to Tailwind. Seemed to me as premature optimization right now.